### PR TITLE
add special type for views of Band, macro to implement banded linear algebra

### PR DESCRIPTION
**Pull request by dlfivefifty**
Now `view(A, band(0))` returns a view of `A`, not `A.data`. 

This enables faster implementations of, for example, `Vector{Float64}(view(A, band(0)))`.

### CODE REVIEW

File: src/generic/linalg.jl [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+    esc(ret)
+end
+
+# add banded linear algebra routines for Typ implementing the BandedMatrix interface
+macro banded_linalg(Typ)
+    ret = quote
+        BandedMatrices.@banded_banded_linalg($Typ, $Typ)
+        # default is to use dense axpy!
+        BandedMatrices.banded_axpy!(a::Number, X::$Typ, Y::AbstractMatrix) =
+                BandedMatrices.banded_dense_axpy!(a, X, Y)
```
Comment (randy3k) [inline src/generic/linalg.jl]: perhaps there is obvious, but why not using `banded_generic_axpy` if the interface has been implemented?

File: src/generic/linalg.jl [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+
+
+banded_generic_matmatmul!(C::AbstractMatrix{T}, tA::Char, tB::Char, A::AbstractMatrix{U}, B::AbstractMatrix{V}) where {T, U, V} = generally_banded_matmatmul!(C, tA, tB, A, B)
+
+
+
+
+macro banded_banded_banded_linalg(Typ1, Typ2, Typ3)
+    ret = quote
+        BandedMatrices.positively_banded_matmatmul!(C::$Typ1{T}, tA::Char, tB::Char, A::$Typ2{T}, B::$Typ3{T}) where {T <: BlasFloat} =
```
Comment (randy3k) [inline src/generic/linalg.jl]: We should only eval this when `leadingdimension` for Typ1, 2, 3 are defined. If they are not defined, this function will dispatch `gbmm` and produce an error.

File: src/generic/linalg.jl [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+
+        Base.LinAlg.A_mul_B!(c::AbstractVector{T}, A::$Typ{U}, b::AbstractVector{V}) where {T, U, V} = BandedMatrices.banded_matvecmul!(c, 'N', A, b)
+        Base.LinAlg.Ac_mul_B!(c::AbstractVector{T}, A::$Typ{U}, b::AbstractVector{V}) where {T, U, V} = BandedMatrices.banded_matvecmul!(c, 'C', A, b)
+        Base.LinAlg.Ac_mul_B!(c::AbstractVector{T}, A::$Typ{U}, b::AbstractVector{V}) where {T, U<:Real, V} = BandedMatrices.banded_matvecmul!(c, 'T', A, b)
+        Base.LinAlg.At_mul_B!(c::AbstractVector{T}, A::$Typ{U}, b::AbstractVector{V}) where {T, U, V} = BandedMatrices.banded_matvecmul!(c, 'T', A, b)
+
+        Base.:*(A::$Typ{U}, b::StridedVector{V}) where {U, V} =
+            Base.LinAlg.A_mul_B!(Vector{promote_type(U, V)}(size(A, 1)), A, b)
+
+        BandedMatrices.positively_banded_matmatmul!(C::StridedMatrix{T}, tA::Char, tB::Char, A::$Typ{T}, B::StridedMatrix{T}) where {T <: BlasFloat} =
```
Comment (randy3k) [inline src/generic/linalg.jl]: same here, need to check `leadingdimension`

File: src/generic/linalg.jl [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+            BandedMatrices._positively_banded_matmatmul!(C, tA, tB, A, B)
+        BandedMatrices.positively_banded_matmatmul!(C::StridedMatrix{T}, tA::Char, tB::Char, A::StridedMatrix{T}, B::$Typ{T}) where {T <: BlasFloat} =
```
Comment (randy3k) [inline src/generic/linalg.jl]: same here, need to check `leadingdimension`

File: src/generic/linalg.jl [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+        end
+
+        function Base.:-(A::UniformScaling, B::$Typ)
+            ret = deepcopy(B)
+            Base.scale!(ret, -1)
+            axpy!(1, A, ret)
+        end
+
+        # use BLAS routine for positively banded BLASBandedMatrix
+        BandedMatrices.positively_banded_matvecmul!(c::StridedVector{T}, tA::Char, A::$Typ{T}, b::StridedVector{T}) where {T <: BandedMatrices.BlasFloat} =
```
Comment (randy3k) [inline src/generic/linalg.jl]: this function should be only defined when the interface `leadingdimension(::Typ{T})` is defined.

**Comment by dlfivefifty** [comment]
I think we should turn BLAS compatible into a trait, a la `Base.HasLength`, and get rid of `BLASBandedMatrix` (see https://github.com/JuliaLang/julia/issues/24383#issuecomment-340282638 for one reason to get rid of `Union` types).  For example:
```julia
struct NotBLAS end
struct IsBLASStridedArray{d} end
const IsBLASStridedMatrix = BLASStridedArray{2}
struct IsBLASBandedMatrix end

blasstructure(::Type{<:AbstractArray}) = NotBLAS()
blasstructure(::Type{SM}) where SM <: StridedMatrix{T} where T <: BlasFloat = IsBLASStridedMatrix()
blasstructure(::Type{BandedMatrix{<:BlasFloat}}) = IsBLASBandedMatrix()
blasstructure(::Type{BandedSubBandedMatrix{<:BlasFloat}}) = IsBLASBandedMatrix()

# if blasstructure returns IsBLASBandedMatrix(), then one is required to override `leadingdimension`
leadingdimension(A::BandedMatrix{<:BlasFloat}) = ...

banded_matmatmul!(C::AbstractMatrix{T}, tA::Char, tB::Char, A::AbstractMatrix{U}, B::AbstractMatrix{V}) where {T, U, V} = banded_matmatmul!(C, tA, tB, A, B, isblasbanded(A), isblasbanded(B))

banded_matmatmul!(C::AbstractMatrix{T}, tA::Char, tB::Char, A::AbstractMatrix{U}, B::AbstractMatrix{V}, blasA, blasB) where {T, U, V} = banded_generic_matmatmul!(C, tA, tB, A, B)

banded_matmatmul!(C::AbstractMatrix{T}, tA::Char, tB::Char, A::AbstractMatrix{U}, B::AbstractMatrix{V}, ::IsBLASBandedMatrix, :IsBLASBandedMatrix) where {T, U, V} = generally_banded_matmatmul!(C, tA, tB, A, B)

```

This does move away from the design of Base, but I would bet that eventually `StridedMatrix` in Base becomes a trait.

**Comment by randy3k** [comment]
Cf: https://github.com/JuliaLang/julia/issues/2345#issuecomment-54537633

**Comment by dlfivefifty** [comment]
I just pushed an update that implements the proposed trait solution. It has simplified the `@banded_linalg` macros.

If I find the time, I might make a similar change in Base and open a pull request, to at least start a discussion about it. I think if Base's design used traits instead of inheritance to determine BLAS structure, it would greatly simplify the `A_*_B!` overloads (and remove the need for the `@banded_linalg` macros completely.

**Pull request merged**